appcues/src/main/java/com/appcues/ui/ModalStateMachineOwner.kt

```diff
@@ -0,0 +1,167 @@
+package com.appcues.ui
+
+import android.view.MotionEvent
+import android.view.View
+import android.view.ViewTreeObserver
+import android.view.ViewTreeObserver.OnDrawListener
+import android.view.ViewTreeObserver.OnGlobalFocusChangeListener
+import android.view.ViewTreeObserver.OnGlobalLayoutListener
+import android.view.ViewTreeObserver.OnScrollChangedListener
+import android.view.ViewTreeObserver.OnTouchModeChangeListener
+import android.view.ViewTreeObserver.OnWindowAttachListener
+import android.view.ViewTreeObserver.OnWindowFocusChangeListener
+import com.appcues.AppcuesCoroutineScope
+import com.appcues.data.model.RenderContext
+import com.appcues.data.model.RenderContext.Modal
+import com.appcues.monitor.AppcuesActivityMonitor
+import com.appcues.statemachine.Action.Retry
+import com.appcues.statemachine.Error.StepError
+import com.appcues.statemachine.StateMachine
+import com.appcues.ui.utils.getParentView
+import kotlinx.coroutines.Dispatchers
+import kotlinx.coroutines.launch
+import java.util.Timer
+import java.util.TimerTask
+import kotlin.concurrent.schedule
+
+internal class ModalStateMachineOwner(
+    override val stateMachine: StateMachine,
+    private val coroutineScope: AppcuesCoroutineScope,
+) : StateMachineOwning {
+
+    companion object {
+        private const val SCROLL_DEBOUNCE_MILLISECONDS = 1_000L
+    }
+
+    override val renderContext: RenderContext = Modal
+
+    // used to attempt retry when items in view change
+    private var viewTreeUpdateHandler: ViewTreeUpdateHandler = ViewTreeUpdateHandler { attemptRetry() }
+    private var uiIdleDebounceTimer: TimerTask? = null
+    private var viewTreeObserver: ViewTreeObserver? = null
+
+    init {
+        // set up observer on the Modal state machine errors to initiate retry
+        coroutineScope.launch(Dispatchers.IO) {
+            stateMachine.errorFlow.collect { error ->
+                // if a recoverable step error is observed, begin attempting retry
+                if (error is StepError && error.recoverable) {
+                    startRetryHandling()
+                }
```

Comment:
we don't want to have retry handling going on all the time in the customer app - as rarely as possible - so this allows it to ONLY happen in special cases of recoverable error, and it will stop (on L81) as soon as a retry is attempted (until another error hits or the screen changes)